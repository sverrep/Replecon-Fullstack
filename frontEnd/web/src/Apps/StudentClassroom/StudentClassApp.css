.student-list{
    float: left;
    width: 30%;
    margin-left: 40px;
    margin-top: 15px;
}

.transfer{
    float: right;
    width: 50%;
}

.wrapper{
    font-family: 'Montserrat', sans-serif;
}

.footer {
    position: fixed;
    padding: 10px 10px 0px 10px;
    bottom: 0;
    width: 100%;
    /* Height of the footer*/ 
    height: 40px;
    background: grey;
}