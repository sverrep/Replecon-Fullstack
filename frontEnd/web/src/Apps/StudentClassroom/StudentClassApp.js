import React from 'react';
import { withRouter } from "react-router-dom";
import axios from 'axios';
import getIP from '../../settings.js';
import './StudentClassApp.css';
import NavBar from '../../Components/navbar/Navbar.js';
import Button from 'react-bootstrap/Button';
import ListGroup from 'react-bootstrap/ListGroup';
import Modal from 'react-bootstrap/Modal';
import Form from 'react-bootstrap/Form';

class StudentClass extends React.Component {
    constructor(props) {
        
        super(props);
        this.state = {
            students: [],
            classroom: '', 
            clicked: {},
            active: false,
            teacher: '',
            class_name: '',
            balance:'',
            name: '',
            value:'',
            sender_id: '',
            recipient_id: '',

        }
        this.alertClicked = this.alertClicked.bind(this)
        this.renderCards = this.renderCards.bind(this)
        this.renderModal = this.renderModal.bind(this)
        this.closeModal = this.closeModal.bind(this)
        this.sendTransfer = this.sendTransfer.bind(this)
    }
    
    componentDidMount(){
        this.getClassStudents()
        this.getStudentBalance()
    }

    getStudentBalance(){
        axios.get(getIP()+'/students/balance/')
                .then(response => {  
                    this.setState({balance:response.data})
                })
                .catch(error => console.log(error))
    }

    createTransaction(){
        for (let i = 0; i <= this.state.students.length-1; i++)
        {
            if (this.state.students[i].name === this.state.name)
            {
                this.setState({ recipient_id: this.state.students[i].id }, () => {
                axios.get(getIP()+'/students/current/')
                .then(response => {
                    this.setState({sender_id: response.data.id}, () => {
                    const payload = { user_id: this.state.recipient_id, amount: this.state.value, recipient: false };
                    const transaction_payload = { recipient_id: this.state.recipient_id, sender_id: this.state.sender_id, category: "Transfer", amount: this.state.value };
                    axios.put(getIP()+'/students/balance/', payload)
                    .then(response => {
                        axios.post(getIP()+'/transactions/', transaction_payload)
                        .then(response => {
                        })
                        .catch(error => console.log(error))
                    })
                    })
                    .catch(error => console.log(error))
                })
                .catch(error => console.log(error))
                this.setState({show:false});
                });
                break;
            }
        }
}

    getClassStudents() {
        axios.get(getIP()+'/students/class_code/')
        .then(response => {
          this.setState({ students: response.data });
          this.setState({ classroom: this.state.students[0].class_code })
          this.getClassroomDetails()
        })
        .catch(error => console.log(error))
    }

    getClassroomDetails(){
        axios.get(getIP()+'/classrooms/')
        .then(response => {
          this.findClassroom(response.data)
    
        })
        .catch(error => console.log(error))
      }
    
      findClassroom(classrooms){
        for (let i = 0; i <= classrooms.length-1; i++)
        {
            
          if (classrooms[i].class_code === this.state.classroom)
          {
            this.setState({ class_name: classrooms[i].class_name });
            axios.get(getIP()+'/teachers/')
            .then(response => {
              for(let j = 0; j <= response.data.length-1; j++)
              {
                if (response.data[j].id === classrooms[i].teacher_id)
                {
                  this.setState({teacher: response.data[j].last_name})
                }
              }
            })
            .catch(error => console.log(error))
          }
        }
      }

    renderCards(item){
        return(
        <ListGroup.Item action onClick={() => this.alertClicked(item)}>
            {item.name}
        </ListGroup.Item>
        )
    }

    alertClicked(item){
        this.setState({clicked: item})
        this.setState({active: true})
        this.setState({name: item.name})
    }

    renderModal(){

        if(this.state.active === true){
            
            return(
            <Modal.Dialog>
                <Modal.Header>
                    <Modal.Title>Money Transfer to: {this.state.clicked.name}</Modal.Title>
                </Modal.Header>

                <Modal.Body>
                    <Form>
                    <Form.Group className="mb-3" controlId="formBasicEmail">
                        
                        <Form.Control onChange={e => this.setState({value: e.target.value})} type="email" placeholder="Enter the amount to send" />
                    </Form.Group>
                    </Form>
                </Modal.Body>

                <Modal.Footer>
                    <Button variant="secondary" onClick={() => this.closeModal()}>Cancel</Button>
                    <Button variant="primary" onClick={() => this.sendTransfer()}>Send</Button>
                </Modal.Footer>
            </Modal.Dialog>
        )
        }
    }
    
    closeModal(){
        this.setState({active:false})
    }

    sendTransfer(){
        console.log(this.state.value)
        this.createTransaction()
    }

    render(){
        return(
            <div>
                <NavBar/>
                
                <h3>{this.state.class_name} - Mr.{this.state.teacher}</h3>
                <div className="student-list">
                    <ListGroup>
                    {this.state.students.map(item => {
                        return(this.renderCards(item))
                    })}
                    </ListGroup>
                    <p>Your classroom: {this.state.classroom}</p>
                </div>
                <div className='transfer'>
                    {this.renderModal()}
                </div>
            </div>
        )
    }
}

export default withRouter(StudentClass)