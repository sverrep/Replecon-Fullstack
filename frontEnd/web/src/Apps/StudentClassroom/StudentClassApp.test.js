import React from 'react';
import StudentClassApp from './StudentClassApp.js'
import axios from 'axios'

jest.mock('axios');
let wrapper
const props = {
    location:{
        state:{
            role: "Student",
            token: 'Token'
        }
    }  
}

describe('Student Class Tests', ()=>{
    
    beforeEach(() => {
        wrapper = shallow(<StudentClassApp.WrappedComponent {...props}/>, { disableLifecycleMethods: true })
    });

    describe('Component did mount', ()=>{
        it('should call the correct functions and retrieve data', async()=>{
            const instance = wrapper.instance()
            
            const user = mockResponseData({id: '1', first_name: 'name'})
            axios.get.mockImplementationOnce(() =>  Promise.resolve(user))

            instance.getStudentBalance = jest.fn()
            instance.getStudentBalance.mockReturnValueOnce()

            instance.getClassStudents = jest.fn()
            instance.getClassStudents.mockReturnValueOnce()

            await instance.componentDidMount()
            expect(instance.state.current_user_id).toBe('1')
            expect(instance.state.current_user_name).toBe('name')

            expect(instance.getStudentBalance).toHaveBeenCalledTimes(1)
            expect(instance.getClassStudents).toHaveBeenCalledTimes(1)

        })
    })

    describe('Student Balance', ()=>{
        it('should retrieve student balance', async()=>{
            const instance = wrapper.instance()
            const user = mockResponseData('100')
            axios.get.mockImplementationOnce(() =>  Promise.resolve(user))

            await instance.getStudentBalance()
            expect(instance.state.balance).toBe('100')

        })  
    })

    describe('Student Class', ()=>{
        it('should get class code and students', async()=>{
            const instance = wrapper.instance()
            const students = mockResponseData([{class_code: '123', name: 'name'}, {class_code: '123', name: 'name'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))

            instance.getClassroomDetails = jest.fn()
            instance.getClassroomDetails.mockReturnValueOnce()

            await instance.getClassStudents()
            expect(instance.state.classroom).toBe('123')
            expect(instance.getClassroomDetails).toHaveBeenCalledTimes(1)
        })

        it('should get classroom details', async()=>{
            const instance = wrapper.instance()
            const classrooms = mockResponseData([{class_code: '123', class_name: 'class name'}, {class_code: '444', class_name: 'class name2'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(classrooms))

            instance.findClassroom = jest.fn()
            instance.findClassroom.mockReturnValueOnce()

            await instance.getClassroomDetails()
            expect(instance.findClassroom).toHaveBeenCalledTimes(1)
        })

        it('should find the classroom', async()=>{
            const instance = wrapper.instance()
            instance.state.classroom = '123'

            const classrooms = [{class_code: '123', class_name: 'class name', teacher_id: '2'}, {class_code: '444', class_name: 'class name2', teacher_id: '3'}]
            const teachers = mockResponseData([{id: '2', last_name: 'last name'}, {id: '23', last_name: 'last name2'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(teachers))
            
            await instance.findClassroom(classrooms)

            expect(instance.state.teacher).toBe('last name')
            expect(axios.get).toHaveBeenCalledTimes(1)
        })

        it('should not find a classroom', async()=>{
            const instance = wrapper.instance()
            instance.state.classroom = '123'

            const classrooms = [{class_code: '123', class_name: 'class name', teacher_id: '2'}, {class_code: '444', class_name: 'class name2', teacher_id: '3'}]
            const teachers = mockResponseData([{id: '234', last_name: 'last name'}, {id: '23', last_name: 'last name2'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(teachers))
            
            await instance.findClassroom(classrooms)
            expect(instance.state.teacher).toBe('')
            expect(axios.get).toHaveBeenCalledTimes(1)
        })
        
    })

    describe('Student Transactions', ()=>{
        it('should pass validate transfer', async ()=>{
            const instance = wrapper.instance()
            instance.state.value = '100'
            instance.state.balance = '200'
            instance.getStudentBalance = jest.fn()
            await instance.validateTransfer()
            expect(instance.state.message).toBe('Money Sent')
        })

        it('should fail validate transfer', async ()=>{
            const instance = wrapper.instance()
            instance.state.value = '300'
            instance.state.balance = '200'
            instance.getStudentBalance = jest.fn()
            await instance.validateTransfer()
            expect(instance.state.message).toBe('You dont have that amount of money')
        })
        //Put is just not registering for some reason wcyd
        it('should create transaction', async()=>{
            const instance = wrapper.instance()
            instance.state.students = [{id:'1'}, {id:'2'}]
            instance.state.value = '100'
            instance.state.clicked = {id: '1'}
            instance.validateTransfer = jest.fn()
            instance.validateTransfer.mockReturnValueOnce(true)

            const teachers = mockResponseData({id: '234'})
            axios.get.mockImplementationOnce(() =>  Promise.resolve(teachers))
            axios.put.mockImplementationOnce(() =>  Promise.resolve())
            axios.post.mockImplementationOnce(() =>  Promise.resolve())

            instance.getStudentBalance = jest.fn()
            instance.getStudentBalance.mockReturnValueOnce()

            await instance.createTransaction()
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.put).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//students/balance/", {"amount": "100", "recipient": false, "user_id": "1"})
            expect(axios.put).toHaveBeenCalledTimes(1)
        })
    })
    
    describe('Studend Class Page Renders and alerts', ()=>{
        it('should show alert', ()=>{
            const instance = wrapper.instance()
            const item = {id: '2', name:'name'}
            instance.state.current_user_id = '1'
            instance.alertClicked(item)
            expect(instance.state.active).toBe(true)
        })

        it('should not show alert', ()=>{
            const instance = wrapper.instance()
            const item = {id: '1', name:'name'}
            instance.state.current_user_id = '1'
            instance.alertClicked(item)
            expect(instance.state.active).toBe(false)
        })

        it('should render page correctly', ()=>{
            expect(wrapper.getElements()).toMatchSnapshot()
        })

        it('should render modal correctly', ()=>{
            const instance = wrapper.instance()
            instance.state.active = true
            expect(instance.renderModal()).toMatchSnapshot()
        })

        it('should render card correctly', ()=>{
            const instance = wrapper.instance()
            const item = {name:'name'}
            expect(instance.renderCards(item)).toMatchSnapshot()
        })

        it('should render alerts correctly', ()=>{
            const instance = wrapper.instance()
            expect(instance.renderAlert('danger', 'message')).toMatchSnapshot()
        })


    })
})

const mockResponseData = jest.fn((data) => {
    if (typeof data === 'string')
    {
        return {
            data: data,
        };
    }
    else
    {
        return {
            data: data,
        };
    }
  });