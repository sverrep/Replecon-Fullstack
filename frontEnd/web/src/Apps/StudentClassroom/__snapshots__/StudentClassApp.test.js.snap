// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Class Tests Studend Class Page Renders and alerts should render alerts correctly 1`] = `undefined`;

exports[`Student Class Tests Studend Class Page Renders and alerts should render card correctly 1`] = `
<ListGroupItem
  action={true}
  onClick={[Function]}
>
  name
</ListGroupItem>
`;

exports[`Student Class Tests Studend Class Page Renders and alerts should render modal correctly 1`] = `
<ModalDialog>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Money Transfer to: 
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form>
      <FormGroup
        className="mb-3"
        controlId="formBasicEmail"
      >
        <FormControl
          onChange={[Function]}
          placeholder="Enter the amount to send"
          type="email"
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="secondary"
    >
      Cancel
    </Button>
    <Button
      active={false}
      className="classroom-button"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Send
    </Button>
  </ModalFooter>
</ModalDialog>
`;

exports[`Student Class Tests Studend Class Page Renders and alerts should render page correctly 1`] = `
Array [
  <div
    className="wrapper"
  >
    <Navbar
      className="teacher-navbar"
      collapseOnSelect={true}
      expand="lg"
      variant="dark"
    >
      <Container
        fluid={false}
      >
        <NavbarBrand
          className="navbar-header"
        >
          Replecon
        </NavbarBrand>
        <Nav
          className="navbar-btns"
          fill={false}
          justify={false}
        >
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Profile",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Profile
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentStore",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Store
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentBank",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Bank
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentClass",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Classroom
            </Link>
          </Button>
        </Nav>
      </Container>
    </Navbar>
    <div
      className="classroom-content"
    >
      <div
        className="title"
      >
        <h3>
          
           - Mr.
          
        </h3>
        <h6>
          Current Balance: 
          
        </h6>
      </div>
      <div
        className="classroom-boxes"
      >
        <div
          className="student-list"
        >
          <ListGroup />
        </div>
        <div
          className="transfer"
        />
      </div>
    </div>
    <div
      className="footer"
    >
      <p>
        Your classroom: 
        
      </p>
    </div>
  </div>,
]
`;
